src/pool/pool_jemalloc.c

```diff
@@ -0,0 +1,301 @@
+
+    return arena_registry_g[arena_ind];
+}
+
+static void *arena_extent_alloc(extent_hooks_t *extent_hooks, void *new_addr,
+                                size_t size, size_t alignment, bool *zero,
+                                bool *commit, unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)new_addr;     // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+
+    void *ptr;
+    enum umf_result_t ret =
+        umfMemoryProviderAlloc(pool->hProvider, size, alignment, &ptr);
+    if (ret != UMF_RESULT_SUCCESS) {
+        return NULL;
+    }
+
+    *zero = true;
+    *commit = true;
+
+    return ptr;
+}
+
+static void arena_extent_destroy(extent_hooks_t *extent_hooks, void *addr,
+                                 size_t size, bool committed,
+                                 unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)committed;    // unused
+
+    jemalloc_memory_pool_t *pool = get_kind_by_arena(arena_ind);
+
+    enum umf_result_t ret = umfMemoryProviderFree(pool->hProvider, addr, size);
+
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr, "umfMemoryProviderFree failed \n");
+    }
+}
+
+static bool arena_extent_dalloc(extent_hooks_t *extent_hooks, void *addr,
+                                size_t size, bool committed,
+                                unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)committed;    // unused
+    (void)arena_ind;    // unused
+
+    return true;
+}
+
+static bool arena_extent_commit(extent_hooks_t *extent_hooks, void *addr,
+                                size_t size, size_t offset, size_t length,
+                                unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    return false;
+}
+
+static bool arena_extent_decommit(extent_hooks_t *extent_hooks, void *addr,
+                                  size_t size, size_t offset, size_t length,
+                                  unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    return true;
+}
+
+static bool arena_extent_purge(extent_hooks_t *extent_hooks, void *addr,
+                               size_t size, size_t offset, size_t length,
+                               unsigned arena_ind) {
+    (void)extent_hooks; // unused
+    (void)addr;         // unused
+    (void)size;         // unused
+    (void)offset;       // unused
+    (void)length;       // unused
+    (void)arena_ind;    // unused
+
+    int err = madvise((char *)addr + offset, length, MADV_DONTNEED);
```

Comment:
I think you should use umfMemoryProviderPurgeForce/umfMemoryProviderPurgeLazy here

Reply:
There is one problem. `arena_extent_purge` is defined as:
```c
static bool arena_extent_purge(extent_hooks_t *extent_hooks, void *addr,
                               size_t size, size_t offset, size_t length, unsigned arena_ind) {
```
and I do not see any way to pass the handle to the memory provider... @igchor do you have any idea how to pass it?

Reply:
Can't we do it similarly to how we obtain provider in `arena_extent_alloc`? (by using `get_kind_by_arena`)

Reply:
Yes, right, I had to be blind yesterday ;-)